### PR TITLE
Secondary Indexes Pt. 1 Episode 1

### PR DESCRIPTION
**Pull request by Hydrocharged**
If a test is missing/should be added, do let me know!

This will definitely fail, the tests catch a few things that the parent branch doesn't have a fix for yet.

### CODE REVIEW

**Comment by Hydrocharged** [comment]
Spoke with @zachmu and we are going to fully disallow tag changes. This means that both of the `Retag column` tests will be removed, and another PR will be submitted (separate from indexes) to block this functionality.

File: bats/index.bats [hunk 1]
```diff
@@ -0,0 +1,1704 @@
...
+    [ "$status" -eq "0" ]
+    [[ "$output" =~ "idx_v1(v1)" ]] || false
+    [[ "$output" =~ "idx_v1v2(v1, v2)" ]] || false
+    run dolt schema show newpk
+    [ "$status" -eq "0" ]
+    [[ "$output" =~ 'INDEX `idx_v1` (`v1`)' ]] || false
+    [[ "$output" =~ 'INDEX `idx_v1v2` (`v1`,`v2`)' ]] || false
+    run dolt index cat newpk idx_v1
+    [ "$status" -eq "0" ]
+    [[ "${lines[1]}" = "| v1 | pk1 |" ]] || false
```
Comment (zachmu) [inline bats/index.bats:561]: Just a spot check is fine here

File: bats/index.bats [hunk 2]
```diff
@@ -0,0 +1,1704 @@
...
+    [[ "${lines[4]}" = "+-----+-----+----+----+" ]] || false
+    # not indexed & not found
+    run dolt sql -q "SELECT * FROM twopk WHERE v1 = 111"
+    [ "$status" -eq "0" ]
+    [[ "${lines[1]}" = "| pk1 | pk2 | v1 | v2 |" ]] || false
+    [[ "${lines[3]}" = "+-----+-----+----+----+" ]] || false
+}
+
+@test "index: SELECT > Full Match" {
+    skip "comparison does not yet use indexing"
```
Comment (zachmu) [inline bats/index.bats:782]: These are just functional tests that should work independent of indexing. I would unskip them. Then if we break them with a bad index implementation, we'll know.

File: bats/index.bats [hunk 3]
```diff
@@ -0,0 +1,1704 @@
...
+    [[ "${lines[5]}" = "+-----+-----+----+----+" ]] || false
+    # not indexed & not found
+    run dolt sql -q "SELECT * FROM twopk WHERE v1 > 111"
+    [ "$status" -eq "0" ]
+    [[ "${lines[1]}" = "| pk1 | pk2 | v1 | v2 |" ]] || false
+    [[ "${lines[3]}" = "+-----+-----+----+----+" ]] || false
+}
+
+@test "index: SELECT < Full Match" {
+    skip "comparison does not yet use indexing"
```
Comment (zachmu) [inline bats/index.bats:851]: Same comment. No reason to skip these, just leave a comment that these tests should work regardless of a working index

File: bats/index.bats [hunk 4]
```diff
@@ -0,0 +1,1704 @@
...
+    [[ "${lines[5]}" = "+-----+-----+----+----+" ]] || false
+    # not indexed & not found
+    run dolt sql -q "SELECT * FROM twopk WHERE v1 < 0"
+    [ "$status" -eq "0" ]
+    [[ "${lines[1]}" = "| pk1 | pk2 | v1 | v2 |" ]] || false
+    [[ "${lines[3]}" = "+-----+-----+----+----+" ]] || false
+}
+
+@test "index: SELECT >= Full Match" {
+    skip "comparison does not yet use indexing"
```
Comment (zachmu) [inline bats/index.bats:921]: Don't skip

File: bats/index.bats [hunk 5]
```diff
@@ -0,0 +1,1704 @@
...
+    [[ "${lines[6]}" = "+-----+-----+----+----+" ]] || false
+    # not indexed & not found
+    run dolt sql -q "SELECT * FROM twopk WHERE v1 >= 111"
+    [ "$status" -eq "0" ]
+    [[ "${lines[1]}" = "| pk1 | pk2 | v1 | v2 |" ]] || false
+    [[ "${lines[3]}" = "+-----+-----+----+----+" ]] || false
+}
+
+@test "index: SELECT <= Full Match" {
+    skip "comparison does not yet use indexing"
```
Comment (zachmu) [inline bats/index.bats:993]: Don't skip

File: bats/index.bats [hunk 6]
```diff
@@ -0,0 +1,1704 @@
...
+    [[ "${lines[7]}" = "| 55 | 2   |" ]] || false
+    [[ "${lines[8]}" = "+----+-----+" ]] || false
+    run dolt sql -q "SELECT * FROM onepk WHERE v2 = 53"
+    [ "$status" -eq "0" ]
+    [[ "${lines[1]}" = "| pk1 | v2 |" ]] || false
+    [[ "${lines[3]}" = "| 5   | 53 |" ]] || false
+    [[ "${lines[4]}" = "+-----+----+" ]] || false
+}
+
+@test "index: Retag column CHANGE COLUMN" {
```
Comment (zachmu) [inline bats/index.bats:1341]: kill this

**Comment by Hydrocharged** [comment]
Manually merged the updated bats tests found here to the main PR https://github.com/liquidata-inc/dolt/pull/495. If desired, the changes can be seen from the commits on that PR.

**Pull request closed without merging**